### ISSUE TITLE
[BUG/Help] assert scale_list.dtype in [torch.half, torch.bfloat16]

### ISSUE THREAD

**Issue by cliuxinxin** [THUDM/ChatGLM2-6B#307]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

torchrun --standalone --nnodes=1 --nproc-per-node=1 ptuning/main.py \
    --do_predict \
    --test_file data/test_1000.json \
    --overwrite_cache \
    --prompt_column content \
    --response_column summary \
    --model_name_or_path ./model/chatglm2-6b \
    --output_dir ./output/base \
    --overwrite_output_dir \
    --per_device_eval_batch_size 1 \
    --quantization_bit 4 \
    --predict_with_generate 

报错如下：

raceback (most recent call last):
  File "/home/pc/inmind/ChatGLM2-6B/ptuning/main.py", line 411, in <module>
    main()
  File "/home/pc/inmind/ChatGLM2-6B/ptuning/main.py", line 377, in main
    predict_results = trainer.predict(predict_dataset, metric_key_prefix="predict", max_length=max_seq_length, do_sample=True, top_p=0.7, temperature=0.95)
  File "/home/pc/inmind/ChatGLM2-6B/ptuning/trainer_seq2seq.py", line 136, in predict
    return super().predict(test_dataset, ignore_keys=ignore_keys, metric_key_prefix=metric_key_prefix)
  File "/usr/local/lib/python3.10/dist-packages/transformers/trainer.py", line 3129, in predict
    output = eval_loop(
  File "/usr/local/lib/python3.10/dist-packages/transformers/trainer.py", line 3245, in evaluation_loop
    loss, logits, labels = self.prediction_step(model, inputs, prediction_loss_only, ignore_keys=ignore_keys)
  File "/home/pc/inmind/ChatGLM2-6B/ptuning/trainer_seq2seq.py", line 202, in prediction_step
    generated_tokens = self.model.generate(**gen_kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/transformers/generation/utils.py", line 1572, in generate
    return self.sample(
  File "/usr/local/lib/python3.10/dist-packages/transformers/generation/utils.py", line 2619, in sample
    outputs = self(
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/pc/.cache/huggingface/modules/transformers_modules/chatglm2-6b/modeling_chatglm.py", line 932, in forward
    transformer_outputs = self.transformer(
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/pc/.cache/huggingface/modules/transformers_modules/chatglm2-6b/modeling_chatglm.py", line 828, in forward
    hidden_states, presents, all_hidden_states, all_self_attentions = self.encoder(
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/pc/.cache/huggingface/modules/transformers_modules/chatglm2-6b/modeling_chatglm.py", line 638, in forward
    layer_ret = layer(
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/pc/.cache/huggingface/modules/transformers_modules/chatglm2-6b/modeling_chatglm.py", line 542, in forward
    attention_output, kv_cache = self.self_attention(
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/pc/.cache/huggingface/modules/transformers_modules/chatglm2-6b/modeling_chatglm.py", line 374, in forward
    mixed_x_layer = self.query_key_value(hidden_states)
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/pc/.cache/huggingface/modules/transformers_modules/chatglm2-6b/quantization.py", line 146, in forward
    output = W8A16Linear.apply(input, self.weight, self.weight_scale, self.weight_bit_width)
  File "/usr/local/lib/python3.10/dist-packages/torch/autograd/function.py", line 506, in apply
    return super().apply(*args, **kwargs)  # type: ignore[misc]
  File "/home/pc/.cache/huggingface/modules/transformers_modules/chatglm2-6b/quantization.py", line 51, in forward
    weight = extract_weight_to_half(quant_w, scale_w, weight_bit_width)
  File "/home/pc/.cache/huggingface/modules/transformers_modules/chatglm2-6b/quantization.py", line 89, in extract_weight_to_half
    assert scale_list.dtype in [torch.half, torch.bfloat16]
AssertionError

### Expected Behavior

本来想预测出数据，但是并没有

### Steps To Reproduce

运行上面的脚本

### Environment

```markdown
- OS:
- Python:
- Transformers:
- PyTorch:
- CUDA Support (`python -c "import torch; print(torch.cuda.is_available())"`) :
```

**Comment by Yehe-code** [comment]
请问你解决了吗，我也遇到了这个问题

**Comment by cliuxinxin** [comment]
没解决

**Comment by MonsterWonder** [comment]
我今天也遇到这个问题了，我们都是想要用ChatGLM2-6B预测测试集的数据，我看源代码里面`pre_seq_len`为空时，作者对全量微调用了`model.float()`，然后我把这里改为了`model.half()`，就能解决该问题了。
`if model_args.pre_seq_len is not None:
        # P-tuning v2
        model = model.half()
        model.transformer.prefix_encoder.float()
    else:
        # Finetune
        model = model.half()`

**Comment by MonsterWonder** [comment]
我今天也遇到这个问题了，我们的目的都是为了用 ChatGLM2-6B 原模型对测试集数据进行测试，我依据报错，去看了源代码，发现 `main.py` 里面有一条处理当嵌入层前缀长度为空时的处理策略，对全量微调模型，作者使用了 `model.float()`，我认为是这里出现了问题，于是我将其改为 `model.half()`，这样就可以正常运行了。
